onadata/apps/viewer/models/parsed_instance.py

```diff
@@ -257,9 +259,16 @@ def update_mongo(self, async=True):
             return False
         else:
             if async:
+                # TODO update self.instance after async save is made
                 update_mongo_instance.apply_async((), {"record": d})
             else:
-                update_mongo_instance(d)
+                success = update_mongo_instance(d)
+                # Only update self.instance is `success` is different from
+                # current_value (`self.instance.is_sync_with_mongo`)
+                if success != self.instance.is_synced_with_mongo:
+                   self.instance.is_synced_with_mongo = success
+                   self.instance.save(update_fields=["is_synced_with_mongo"], force=True)
```

Comment:
`update_fields` is great as implemented by Django, but `Instance.save()` doesn't respect it very well. Ideas:
1. Skip the [labor-intensive](https://github.com/kobotoolbox/kobocat/blob/fb334b3151415855181df256658845b9a423b7cb/onadata/apps/logger/models/instance.py#L376-L380) stuff in `Instance.save()` if
    * (simple) `save()` gets passed a special argument, e.g. `content_update=False`, or
    * (perhaps too convoluted) `update_fields` contains only certain "safe" fields that don't require `_set_geom()`, `_set_json()`, and friends, or
2. Call `update()` on a single-item query set instead of `save()` on the `Instance` object. It's kind of inelegant, but I occasionally use the pattern `Thing.objects.filter(pk=thing_to_update.pk).update(field_to_update='new value')` to bypass troublesome `save()` overrides and write to the database efficiently.

Reply:
Go for solution #2